// @sap-cloud-sdk/connectivity/dist/scp-cf/destination/service-binding-to-destination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serviceToDestinationTransformers = void 0;
const token_accessor_1 = require("../token-accessor");
const jwt_1 = require("../jwt");
/**
 * @internal
 */
exports.serviceToDestinationTransformers = {
    'business-logging': businessLoggingBindingToDestination,
    's4-hana-cloud': xfS4hanaCloudBindingToDestination,
    destination: destinationBindingToDestination,
    'saas-registry': saasRegistryBindingToDestination,
    workflow: workflowBindingToDestination,
    'service-manager': serviceManagerBindingToDestination
};
async function serviceManagerBindingToDestination(serviceBinding, options) {
    const service = {
        ...serviceBinding,
        tags: serviceBinding.tags,
        label: 'service-manager',
        credentials: { ...serviceBinding.credentials }
    };
    const token = await (0, token_accessor_1.serviceToken)(service, options);
    return buildClientCredentialsDestination(token, serviceBinding.credentials.sm_url, serviceBinding.name);
}
async function destinationBindingToDestination(serviceBinding, options) {
    const service = {
        ...serviceBinding,
        tags: serviceBinding.tags,
        label: 'destination',
        credentials: { ...serviceBinding.credentials }
    };
    const token = await (0, token_accessor_1.serviceToken)(service, options);
    return buildClientCredentialsDestination(token, serviceBinding.credentials.uri, serviceBinding.name);
}
async function saasRegistryBindingToDestination(serviceBinding, options) {
    const service = {
        ...serviceBinding,
        tags: serviceBinding.tags,
        label: 'saas-registry',
        credentials: { ...serviceBinding.credentials }
    };
    const token = await (0, token_accessor_1.serviceToken)(service, options);
    return buildClientCredentialsDestination(token, serviceBinding.credentials['saas_registry_url'], serviceBinding.name);
}
async function businessLoggingBindingToDestination(serviceBinding, options) {
    const service = {
        ...serviceBinding,
        tags: serviceBinding.tags,
        label: 'business-logging',
        credentials: { ...serviceBinding.credentials.uaa }
    };
    const token = await (0, token_accessor_1.serviceToken)(service, options);
    return buildClientCredentialsDestination(token, serviceBinding.credentials.writeUrl, serviceBinding.name);
}
async function workflowBindingToDestination(serviceBinding, options) {
    const service = {
        ...serviceBinding,
        tags: serviceBinding.tags,
        label: 'workflow',
        credentials: { ...serviceBinding.credentials.uaa }
    };
    const token = await (0, token_accessor_1.serviceToken)(service, options);
    return buildClientCredentialsDestination(token, serviceBinding.credentials.endpoints.workflow_odata_url, serviceBinding.name);
}
async function xfS4hanaCloudBindingToDestination(serviceBinding) {
    return {
        url: serviceBinding.credentials.URL,
        authentication: 'BasicAuthentication',
        username: serviceBinding.credentials.User,
        password: serviceBinding.credentials.Password
    };
}
function buildClientCredentialsDestination(token, url, name) {
    const expiresIn = Math.floor(((0, jwt_1.decodeJwt)(token).exp * 1000 - Date.now()) / 1000).toString(10);
    return {
        url,
        name,
        authentication: 'OAuth2ClientCredentials',
        authTokens: [
            {
                value: token,
                type: 'bearer',
                expiresIn,
                http_header: { key: 'Authorization', value: `Bearer ${token}` },
                error: null
            }
        ]
    };
}
